draft-ietf-quic-recovery.md

```diff
@@ -398,8 +417,82 @@ A packet sent on an RTO alarm MUST NOT be blocked by the sender's congestion
+losing the client's Initial, but receiving the 0-RTT packets.  Below are
+three different optimizations in increasing complexity that minimize
+handshake latency.
+
+### EMPTY_ACK frames
+
+The EMPTY_ACK frame indicates a packet for the connection was received,
+but it could not be processed, because the decryption keys are not yet
+available. EMPTY_ACK enables faster recovery of lost Initial
+and Handshake packets when the only other outstanding packets are
+undecryptable.
+
+The receiver SHOULD send an EMPTY_ACK frame soon (e.g., 1ms) after
+undecryptable packets are received, even if those received packets are
+not buffered for later decryption.  The small delay allows for cases when
+0-RTT packets are reordered in front of the Initial, which is not uncommon
+on networks that prioritize small packets.  The receiver should limit the
+number of EMPTY_ACK frames sent to one per packet number space per RTT.
+If no RTT is known, only one per encryption level should be sent.
+
+When an EMPTY_ACK frame is received, a sender should immediately
+retransmit the missing handshake packet(s) as though the handshake timer
+fired and re-arm the handshake timer when the handshake packets are sent.
+If the missing handshake data was timer retransmitted after the packets
+that triggered the empty ack, then the empty ack should be ignored.
+
+An EMPTY_ACK frame does not acknowledge a new packet, and in cases
+when multiple packets are outstanding, the RTT signal is ambiguous,
+so it should not be used like an RTT signal from a newly acknowledged
+packet.  It MAY change the connection’s default RTT if no RTT measurements
+have been taken.
+
+### Coalesced Packets
+
+Despite loss recovery being separate for different packet number spaces,
+the ability to send a coalesced packet enables faster recovery with small,
+and sometimes no overhead.  The acknowledgement of a coalesced packet
+allows QUIC recovery to use early retransmit to determine if any prior
+packets in that space were lost without waiting for timeouts.
+
+This optimization is particularly useful when:
+
+ * Sending the client’s Initial, which must be padded to a full
+   sized packet, so the datagram typically has extra space to retransmit
+   some outstanding 0-RTT data.
+ * The clients sends 1-RTT data soon after the final TLS flight
+   (containing the client Finished) and can proactively retransmit the
+   final client flight with one or more 1-RTT packets.
+
+### Implicit Acknowledgements
+
+Handshake data may be cancelled when packets at a higher encryption
+level are processed, as this demonstrates the peer has received the
+handshake data at the prior encryption level.
+
+In particular:
+
+ * Processing data in a Handshake packet indicates the Initial
+   packet(s) have been delivered.
+ * A Server processing 1-RTT packets indicates all CRYPTO_HS data in
+   Handshake packets has been delivered.
+ * Processing 0-RTT packets does not indicate the peer has received
+   any handshake data.
+
```

Comment:
This need a whole lot more explaining. Consider a server that sends handshake packets containing its first flight, including server finished, and then sends 1-RTT packets, e.g. ACK and response to 0-RTT packets. The client sends its second flight, including client finished, and then sends 1-RTT packet. Should the server acknowledge refuse to process and acknowledge the 1-RTT packet until the handshake packet for the client finished has been received and processed? Is that the intent?